packages/astro/src/cli/check/index.ts

```diff
@@ -1,101 +1,360 @@
+	 * 2. Optionally, traverse the file system for `.astro` files and saves their paths.
+	 * 3. Get diagnostics for said files and print the result in console.
+	 *
+	 * @param openDocuments Whether the operation should open all `.astro` files
+	 * @private
+	 */
+	async #checkAllFiles(openDocuments: boolean): Promise<CheckResult> {
+		const processExit = await this.#syncCli(this.#settings, {
+			logging: this.#logging,
+			fs: this.#fs,
+		});
+		// early exit on sync failure
+		if (processExit === 1) return processExit;
+
+		let spinner = ora(
+			` Getting diagnostics for Astro files in ${fileURLToPath(this.#settings.config.root)}…`
+		).start();
+
+		if (openDocuments) {
+			await this.#openDocuments();
+		}
+
+		let diagnostics = await this.#diagnosticsChecker.getDiagnostics();
+
+		spinner.succeed();
+
+		let brokenDownDiagnostics = this.#breakDownDiagnostics(diagnostics);
+		this.#logDiagnosticsSeverity(brokenDownDiagnostics);
+		return brokenDownDiagnostics.errors > 0
+			? CheckResult.ExitWithError
+			: CheckResult.ExitWithSuccess;
+	}
+
+	#checkForDiagnostics() {
+		clearTimeout(this.#updateDiagnostics);
+		// @ematipico: I am not sure of `setTimeout`. I would rather use a debounce but let's see if this works.
+		// Inspiration from `svelte-check`.
+		this.#updateDiagnostics = setTimeout(async () => await this.#checkAllFiles(false), 500);
+	}
+
+	/**
+	 * This function is responsible to attach events to the server watcher
+	 */
+	async #watch() {
+		const { default: chokidar } = await import('chokidar');
+		this.#watcher = chokidar.watch(
+			join(fileURLToPath(this.#settings.config.root), ASTRO_GLOB_PATTERN),
+			{
+				ignored: ['**/node_modules/**'],
+				ignoreInitial: true,
 			}
+		);
+
+		this.#watcher.on('add', (file) => {
+			this.#addDocument(file);
+			this.#filesCount += 1;
+			this.#checkForDiagnostics();
 		});
-	});
+		this.#watcher.on('change', (file) => {
+			this.#addDocument(file);
+			this.#checkForDiagnostics();
+		});
+		this.#watcher.on('unlink', (file) => {
+			this.#diagnosticsChecker.removeDocument(file);
+			this.#filesCount -= 1;
+			this.#checkForDiagnostics();
+		});
+	}
 
-	console.log(
-		[
-			bold(`Result (${filesCount} file${filesCount === 1 ? '' : 's'}): `),
-			bold(red(`${result.errors} ${result.errors === 1 ? 'error' : 'errors'}`)),
-			bold(yellow(`${result.warnings} ${result.warnings === 1 ? 'warning' : 'warnings'}`)),
-			dim(`${result.hints} ${result.hints === 1 ? 'hint' : 'hints'}\n`),
-		].join(`\n${dim('-')} `)
-	);
+	/**
+	 * Add a document to the diagnostics checker
+	 * @param filePath Path to the file
+	 */
+	#addDocument(filePath: string) {
+		const text = fs.readFileSync(filePath, 'utf-8');
+		this.#diagnosticsChecker.upsertDocument({
+			uri: pathToFileURL(filePath).toString(),
+			text,
+		});
+	}
+
+	/**
+	 * Logs the result of the various diagnostics
+	 *
+	 * @param {Readonly<DiagnosticResult>} result Result emitted by AstroChecker.#breakDownDiagnostics
```

Comment:
I think we can remove the duplicated types here too:
```suggestion
	 * @param result Result emitted by AstroChecker.#breakDownDiagnostics
```